src/AWS.Deploy.ServerMode.Client/ServerModeSession.cs

```diff
@@ -0,0 +1,152 @@
+
+                var command = $"dotnet aws server-mode --port {port} --parent-pid {currentProcessId} --encryption-keyinfo-stdin";
+                var startServerTask = RunCommand(command, keyInfoStdin);
+
+                var baseUrl = $"http://localhost:{port}";
+                var isServerAvailableTask = IsServerAvailable(baseUrl, cancellationToken);
+
+                if (isServerAvailableTask == await Task.WhenAny(startServerTask, isServerAvailableTask))
+                {
+                    // If the server health check succeeded
+                    // return the baseUrl
+                    if (isServerAvailableTask.Result)
+                    {
+                        return new ServerModeStartInfo(baseUrl, aes);
+                    }
+
+                    // The server timed out, this isn't a transient error, therefore, we throw
+                    throw new InternalServerModeException($"\"{command}\" failed for unknown reason.");
+                }
+
+                // For -100 errors, we want to check all the ports in the configured port range
+                // If the error code other than -100, this is an unexpected exit code.
+                if (startServerTask.Result != TCP_PORT_ERROR)
+                {
+                    throw new InternalServerModeException($"\"{command}\" failed for unknown reason.");
+                }
+            }
+
+            throw new PortUnavailableException($"Free port unavailable in {_startPort}-{_endPort} range.");
+        }
+
+        private Task<bool> IsServerAvailable(string baseUrl, CancellationToken cancellationToken)
+        {
+            var healthUrl = $"{baseUrl}/api/v1/health";
+            return WaitUntilHelper.WaitUntilSuccessStatusCode(healthUrl, TimeSpan.FromMilliseconds(100), TimeSpan.FromMinutes(10), cancellationToken);
+        }
+
+        private async Task<int> RunCommand(string command, params string[] stdIn)
+        {
+            var arguments = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? $"/c {command}" : $"-c \"{command}\"";
+
+            var processStartInfo = new ProcessStartInfo
+            {
+                FileName = GetSystemShell(),
+                Arguments = arguments,
+                UseShellExecute = false, // UseShellExecute must be false in allow redirection of StdIn.
+                RedirectStandardInput = true,
+                CreateNoWindow = !_diagnosticLoggingEnabled, // It allows displaying stdout and stderr on the screen.
+            };
+
+            var process = Process.Start(processStartInfo);
+            if (process == null)
+            {
+                throw new InvalidOperationException();
+            }
+
+            foreach (var line in stdIn)
+            {
+                await process.StandardInput.WriteLineAsync(line);
+            }
+
+            process.WaitForExit(-1);
+
+            return await Task.FromResult(process.ExitCode);
```

Comment:
These two lines may need to get wrapped in a `Task.Run` call. `WaitForExit` appears to lock things up for me (the other task that checks for a successful status code does not proceed to run).

Reply:
is  this because you are calling the method from main thread?

Reply:
I am, however the main UI thread doesn't get locked up as things are still responsive. The Tasks that are created in ServerModeSession::Start are what appear to stop running. The call `isServerAvailableTask == await Task.WhenAny(startServerTask, isServerAvailableTask)` never completes.

Reply:
Can you try with latest changes pushed, I tested in an WPF app and it is working expectedly.